x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/permission/ClusterPermissionTests.java

```diff
@@ -0,0 +1,219 @@
+        }
+        {
+            final ClusterPermission.Builder builder = ClusterPermission.builder();
+            ClusterPrivilegeResolver.MANAGE_SECURITY.buildPermission(builder);
+            ClusterPrivilegeResolver.MANAGE_ILM.buildPermission(builder);
+            final MockConfigurableClusterPrivilege mockConfigurableClusterPrivilege1 =
+                new MockConfigurableClusterPrivilege(r -> r == mockTransportRequest);
+            final MockConfigurableClusterPrivilege mockConfigurableClusterPrivilege2 =
+                new MockConfigurableClusterPrivilege(r -> false);
+            mockConfigurableClusterPrivilege1.buildPermission(builder);
+            mockConfigurableClusterPrivilege2.buildPermission(builder);
+
+            ClusterPermission clusterPermission = builder.build();
+            ClusterPermission impliedBy = ClusterPermission.builder().build();
+            assertThat(clusterPermission.implies(impliedBy), is(false));
+        }
+    }
+
+    public void testClusterPermissionImplies() {
+        final ClusterPermission.Builder builder = ClusterPermission.builder();
+        ClusterPrivilegeResolver.MANAGE_SECURITY.buildPermission(builder);
+        ClusterPrivilegeResolver.MANAGE_ILM.buildPermission(builder);
+        final MockConfigurableClusterPrivilege mockConfigurableClusterPrivilege1 =
+            new MockConfigurableClusterPrivilege(r -> r == mockTransportRequest);
+        final MockConfigurableClusterPrivilege mockConfigurableClusterPrivilege2 =
+            new MockConfigurableClusterPrivilege(r -> false);
+        mockConfigurableClusterPrivilege1.buildPermission(builder);
+        mockConfigurableClusterPrivilege2.buildPermission(builder);
+
+        {
+            final TransportRequest mockTransportRequest3 = Mockito.mock(TransportRequest.class);
+            final MockConfigurableClusterPrivilege mockConfigurableClusterPrivilege3 =
+                new MockConfigurableClusterPrivilege(r -> r == mockTransportRequest3);
+            ClusterPermission clusterPermission = builder.build();
+            ClusterPermission impliedBy = mockConfigurableClusterPrivilege3.buildPermission(ClusterPermission.builder()).build();
+            assertThat(clusterPermission.implies(impliedBy), is(false));
+        }
+        {
+            ClusterPermission clusterPermission = ClusterPrivilegeResolver.ALL.buildPermission(ClusterPermission.builder()).build();
+            ClusterPermission impliedBy = ClusterPrivilegeResolver.ALL.buildPermission(ClusterPermission.builder()).build();
+            assertThat(clusterPermission.implies(impliedBy), is(true));
+
+            clusterPermission = ClusterPrivilegeResolver.ALL.buildPermission(ClusterPermission.builder()).build();
+            impliedBy = ClusterPrivilegeResolver.MANAGE_ILM.buildPermission(ClusterPermission.builder()).build();
+            assertThat(clusterPermission.implies(impliedBy), is(true));
```

Comment:
from the last two tests I think I got `implies` wrong? I am thinking that something which is implied (the argument) is subsumed by the object, not vice-versa?